File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/handling/PacketActionHandler.java
package com.gmail.marcosav2010.communicator.packet.handling;

import com.gmail.marcosav2010.communicator.packet.Packet;
import com.gmail.marcosav2010.connection.Connection;
import com.gmail.marcosav2010.logger.ILog;
import com.gmail.marcosav2010.logger.Log;
import com.gmail.marcosav2010.tasker.TaskOwner;
import com.gmail.marcosav2010.tasker.Tasker;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

/**
 * This class handles the action is done when a packet is received by remote
 * peer.
 *
 * @author Marcos
 */
public class PacketActionHandler {

    private static final long MAX_TIMEOUT = 10L;

    private final ILog log;

    private final Map<Long, PacketAction> pendingActions = new ConcurrentHashMap<>();

    public PacketActionHandler(Connection connection) {
        log = new Log(connection, "PacketActionHandler");
    }

    public boolean isPending(long id) {
        return pendingActions.containsKey(id);
    }

    public void handleRespose(long id) {
        if (!isPending(id))
            return;

        PacketAction pa = pendingActions.remove(id);

        try {
            pa.onReceive();
        } catch (Exception e) {
            log.log(e, "There was an error while handling action:\n\tID: " + id + "\n\tPacket: "
                    + pa.getType().getName() + "\n\tStacktrace: ");
        }
    }

    public void handleSend(TaskOwner owner, long id, Packet packet, Runnable action, Runnable onTimeOut,
                           long expireTimeout, TimeUnit timeUnit) {
        PacketAction pa = new PacketAction(action, onTimeOut);
        pa.setType(packet.getClass());
        pendingActions.put(id, pa);
        if (timeUnit == null || expireTimeout < 0) {
            expireTimeout = MAX_TIMEOUT;
            timeUnit = TimeUnit.SECONDS;
        }

        Tasker.getInstance().schedule(owner, () -> onExpire(id), expireTimeout, timeUnit);
    }

    public void onExpire(long id) {
        if (!isPending(id))
            return;

        PacketAction pa = pendingActions.remove(id);

        try {
            pa.onTimeOut();
        } catch (Exception e) {
            log.log(e, "There was an error while handling time out action:\n\tID: " + id + "\n\tPacket: "
                    + pa.getType().getName() + "\n\tStacktrace: ");
        }
    }
}

File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/handling/PacketActionHandler.kt
package com.gmail.marcosav2010.communicator.packet.handling

import com.gmail.marcosav2010.communicator.packet.Packet
import com.gmail.marcosav2010.connection.Connection
import com.gmail.marcosav2010.logger.ILog
import com.gmail.marcosav2010.logger.Log
import com.gmail.marcosav2010.tasker.TaskOwner
import com.gmail.marcosav2010.tasker.Tasker
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/**
 * This class handles the action is done when a packet is received by remote
 * peer.
 *
 * @author Marcos
 */
class PacketActionHandler(connection: Connection) {

    companion object {
        private const val MAX_TIMEOUT_SECONDS = 10L
    }

    private val log: ILog = Log(connection, "PAH")
    private val pendingActions: MutableMap<Long, PacketAction> = ConcurrentHashMap()

    private fun isPending(id: Long): Boolean = pendingActions.containsKey(id)

    fun handleResponse(id: Long) {
        if (!isPending(id)) return
        val pa = pendingActions.remove(id)
        try {
            pa!!.onReceive()
        } catch (e: Exception) {
            log.log(e, """There was an error while handling action:
	ID: $id
	Packet: ${pa!!.type.name}
	Stacktrace: """)
        }
    }

    fun handleSend(owner: TaskOwner,
                   id: Long,
                   packet: Packet,
                   action: () -> Unit,
                   onTimeOut: (() -> Unit)?,
                   expireTimeout: Long,
                   timeUnit: TimeUnit) {

        check(timeUnit.toSeconds(expireTimeout) <= MAX_TIMEOUT_SECONDS) {
            "Expire timeout must be lower or equal than $MAX_TIMEOUT_SECONDS seconds"
        }

        val pa = PacketAction(action, onTimeOut, packet.javaClass)
        pendingActions[id] = pa

        Tasker.schedule(owner, { onExpire(id) }, expireTimeout, timeUnit)
    }

    private fun onExpire(id: Long) {
        if (!isPending(id)) return

        val pa = pendingActions.remove(id)
        try {
            pa!!.onTimeOut()
        } catch (e: Exception) {
            log.log(e, """There was an error while handling time out action:
	ID: $id
	Packet: ${pa!!.type.name}
	Stacktrace: """)
        }
    }
}